### PR TITLE
negcache improvements

### LINKED ISSUE #2793: improve negative caching
port #2770 to master

### PR DESCRIPTION
**Pull request by pieterlexis**
Closes #2793 by porting the changes made in #2770 to master.

### CODE REVIEW

File: pdns/dnsbackend.cc [hunk 1]
```diff
@@ -33,15 +33,18 @@
...
-    if( best_match_len >= (int)subdomain.toString().length() )
+    if( best_match_len >= (int)subdomain.toString().length() && p->qtype != QType::DS )
       break;
 
-    if( this->getSOA( subdomain, *sd, p ) ) {
+    map<DNSName,int>::iterator it = negCacheMap.find(subdomain);
+    bool negCached = ( it != negCacheMap.end() && it->second == 1 );
+
+    if(! negCached && this->getSOA( subdomain, *sd, p ) ) {
       sd->qname = subdomain;
```
Comment (mind04) [inline pdns/dnsbackend.cc]: +      if(found) // Second SOA found, we are done
+        return true;

File: pdns/ueberbackend.cc [hunk 2]
```diff
@@ -297,6 +307,23 @@ bool UeberBackend::getAuth(DNSPacket *p, SOAData *sd, const DNSName &target)
         if( best_match_len == (int)target.countLabels() )
             goto auth_found;
```
Comment (mind04) [inline pdns/ueberbackend.cc]: the four lines above are removed in  #2770

**Comment by Habbie** [comment]
@mind04 your thoughts please

**Comment by mind04** [comment]
LGTM

File: regression-tests/tests/bind-add-zone/expected_result.bind [hunk 3]
```diff
@@ -30,7 +30,7 @@ Reply to question for qname='ns1.addzone.com.', qtype=A
 Rcode: 0 (No Error), RD: 0, QR: 1, TC: 0, AA: 1, opcode: 0
 Reply to question for qname='ns1.test.com.', qtype=A
 Loaded zone addzone.com from zones/addzone.com
-0
+1
```
Comment (Habbie) [inline regression-tests/tests/bind-add-zone/expected_result.bind]: Mental note: wonder if this 0/1 change is related to #3078  - probably not

**Pull request merged**